lib/services/local/access_list.go

```diff
@@ -259,67 +278,87 @@ func (a *AccessListService) DeleteAllAccessListMembers(ctx context.Context) erro
 
-			for _, member := range members {
-				// If the member is not in the members map (request), delete it.
-				if _, ok := membersMap[member.GetName()]; !ok {
-					err = a.memberService.WithPrefix(accessList.GetName()).DeleteResource(ctx, member.GetName())
-					if err != nil {
-						return trace.Wrap(err)
-					}
-				} else {
-					// Compare members and update if necessary.
-					if !cmp.Equal(member, membersMap[member.GetName()]) {
-						// Update the member.
-						err = a.memberService.WithPrefix(accessList.GetName()).UpsertResource(ctx, membersMap[member.GetName()])
+			var (
+				members      []*accesslist.AccessListMember
+				membersToken string
+				err          error
+			)
+
+			for {
+				// List all members for the access list.
+				members, membersToken, err = a.memberService.WithPrefix(accessList.GetName()).ListResources(ctx, 0 /* default size */, membersToken)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				for _, member := range members {
+					// If the member is not in the members map (request), delete it.
+					if _, ok := membersMap[member.GetName()]; !ok {
+						err = a.memberService.WithPrefix(accessList.GetName()).DeleteResource(ctx, member.GetName())
 						if err != nil {
 							return trace.Wrap(err)
 						}
+					} else {
+						// Compare members and update if necessary.
+						if !cmp.Equal(member, membersMap[member.GetName()]) {
+							// Update the member.
+							err = a.memberService.WithPrefix(accessList.GetName()).UpsertResource(ctx, membersMap[member.GetName()])
+							if err != nil {
+								return trace.Wrap(err)
+							}
+						}
 					}
+
+					// Remove the member from the map.
+					delete(membersMap, member.GetName())
 				}
 
-				// Remove the member from the map.
-				delete(membersMap, member.GetName())
+				if membersToken == "" {
+					break
+				}
 			}
 
-			if membersToken == "" {
-				break
+			// Add any remaining members to the access list.
+			for _, member := range membersMap {
+				err = a.memberService.WithPrefix(accessList.GetName()).UpsertResource(ctx, member)
+				if err != nil {
+					return trace.Wrap(err)
+				}
 			}
-		}
 
-		// Add any remaining members to the access list.
-		for _, member := range membersMap {
-			err = a.memberService.WithPrefix(accessList.GetName()).UpsertResource(ctx, member)
-			if err != nil {
-				return trace.Wrap(err)
-			}
+			return trace.Wrap(a.service.UpsertResource(ctx, accessList))
+		})
+	}
+
+	var completeUpsertFn func() error
+	if feature := modules.GetModules().Features(); !feature.IdentityGovernance {
+		completeUpsertFn = func() error {
+			return a.service.RunWhileLocked(ctx, "createAccessListWithMembersLimitLock", accessListLockTTL, func(ctx context.Context, _ backend.Backend) error {
+				if err := a.VerifyAccessListCreateLimit(ctx, accessList.GetName()); err != nil {
+					return trace.Wrap(err)
+				}
+				return trace.Wrap(upsertWithLockFn())
+			})
 		}
+	} else {
+		completeUpsertFn = upsertWithLockFn
```

Comment:
The same as above. I look at this function and see a lot of code that may not be executed as here we swap the function that will be called. Can we refactor the code and possibly move the code out, to a separate function to make this part cleaner?